### **PROJECT 2**
Released on 11-July-2022

Due Date for Submission 20-July-2022 at 23:59 hrs IST


The Student has been provided with a dataset for this Project. They will have to use the tools and techniques learnt in this course and answer the below questions for their submission.


The Format of submission is as follows : 

A .csv file with Question numbers and answers written.

Images for specific questions have to be uploaded as a response in the given form


Please check the [Sample Submission file](https://drive.google.com/file/d/1CYSjy8HJCmZo24_yU08bKaO-Rvs4-Mmo/view)

Also check the [Image Upload Link](https://docs.google.com/forms/d/1Kx0LcsKXQoRGSncH-Xb1ExA39N7sIy0wzE1MctYgyvA/viewform?edit_requested=true)

We will be evaluating both the .csv file and the published links. Any formatting issue in the .csv file will result in zero marks for the relevant question(s).





### **FILES :** 


**ratings.csv**  contains ratings sorted by time. Ratings go from one to five. Both book IDs and user IDs are contiguous. For books, they are 1-10000, for users, 1-53424.

**to_read.csv** provides IDs of the books marked "to read" by each user, as user_id,book_id pairs, sorted by time.

**books.csv** has metadata for each book (goodreads IDs, authors, title, average rating, etc.). The metadata has been extracted from goodreads XML files.

**book_tags.csv** contains tags/shelves/genres assigned by users to books. Tags in this file are represented by their IDs. Each book_id  has multiple tag_id.The field "count" denotes ‘user records’ (the number of users tagged the given tag_id with the goodreads_book_id).



### **NOTES :**

*Answering:*

* Dataset to be referred will be mentioned in the square brackets for each question.eg [books.csv]

* String answers (“S”)   will be evaluated against entries from the column “title” in [books.csv]

*Data Cleaning:*

* For Q1 Find the value from the original csv file [books.csv]. For remaining questions  remove null values in 'original_title'.

* Ensure for  Q14 null values are removed from  'original_title' to perform sentiment analysis.


### **QUESTIONS :**



In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 

In [2]:
pd.set_option('float_format', '{:f}'.format)

In [3]:
book_tags = pd.read_csv('DATASET/book_tags.csv')
book_tags.head()

goodreads_book_id  tag_id  count
0                 50    8717   2212
1                 50   33114   2121
2                 50   11743   1587
3                 50    1691    934
4                 50   11557    890

In [4]:
book_tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55800 entries, 0 to 55799
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   goodreads_book_id  55800 non-null  int64
 1   tag_id             55800 non-null  int64
 2   count              55800 non-null  int64
dtypes: int64(3)
memory usage: 1.3 MB


In [5]:
books = pd.read_csv('DATASET/books.csv')
books.head()

book_id  goodreads_book_id  best_book_id  work_id  books_count       isbn  \
0       36               3636          3636  2543234          192  385732554   
1       37             100915        100915  4790821          474   60764899   
2       50              30119         30119    30518           45   60513039   
3       59              24178         24178   987048          180   64410935   
4       77              38709         38709  1679789          138  439244196   

                isbn13                                     authors  \
0 9780385732550.000000                                  Lois Lowry   
1 9780060764890.000000                                  C.S. Lewis   
2 9780060513030.000000                            Shel Silverstein   
3 9780064410940.000000  E.B. White, Garth Williams, Rosemary Wells   
4 9780439244190.000000                  Louis Sachar, Louis Sachar   

   original_publication_year  \
0                1993.000000   
1                1950.000000   
2                1974.000000   
3                1952.000000   
4                1998.000000   

                                      original_title  ... work_ratings_count  \
0                                          The Giver  ...            1345445   
1               The Lion, the Witch and the Wardrobe  ...            1584884   
2  Where the Sidewalk Ends: The Poems and Drawing...  ...            1023781   
3                                    Charlotte's Web  ...            1084842   
4                                              Holes  ...             764637   

  work_text_reviews_count  ratings_1  ratings_2  ratings_3  ratings_4  \
0                   54084      26497      59652     225326     448691   
1                   15186      19309      55542     262038     513366   
2                    9234      16590      30792     139024     293222   
3                   13044      18450      41124     191415     346090   
4                   15832      15495      43786     176946     269153   

   ratings_5                                          image_url  \
0     585279  https://images.gr-assets.com/books/1342493368m...   
1     734629  https://images.gr-assets.com/books/1353029077m...   
2     544153  https://images.gr-assets.com/books/1168052448m...   
3     487763  https://images.gr-assets.com/books/1439632243m...   
4     259257  https://images.gr-assets.com/books/1327781893m...   

                                     small_image_url  NonEnglish  
0  https://images.gr-assets.com/books/1342493368s...           0  
1  https://images.gr-assets.com/books/1353029077s...           0  
2  https://images.gr-assets.com/books/1168052448s...           0  
3  https://images.gr-assets.com/books/1439632243s...           0  
4  https://images.gr-assets.com/books/1327781893s...           0  

[5 rows x 24 columns]

In [6]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 556 entries, 0 to 555
Data columns (total 24 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   book_id                    556 non-null    int64  
 1   goodreads_book_id          556 non-null    int64  
 2   best_book_id               556 non-null    int64  
 3   work_id                    556 non-null    int64  
 4   books_count                556 non-null    int64  
 5   isbn                       552 non-null    object 
 6   isbn13                     556 non-null    float64
 7   authors                    556 non-null    object 
 8   original_publication_year  552 non-null    float64
 9   original_title             517 non-null    object 
 10  title                      556 non-null    object 
 11  language_code              476 non-null    object 
 12  average_rating             556 non-null    float64
 13  ratings_count              556 non-null    int64  

In [7]:
ratings = pd.read_csv('Dataset/ratings.csv')
ratings.head()

user_id  book_id  rating
0        4      388       4
1        4       36       5
2        4      102       5
3        4       50       4
4        4      158       4

In [8]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 461029 entries, 0 to 461028
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype
---  ------   --------------   -----
 0   user_id  461029 non-null  int64
 1   book_id  461029 non-null  int64
 2   rating   461029 non-null  int64
dtypes: int64(3)
memory usage: 10.6 MB


In [9]:
toread = pd.read_csv('Dataset/toread.csv')
toread.head()

user_id  book_id
0      506      241
1      504      219
2      564       36
3      756      621
4      589     1977

In [10]:
toread.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28711 entries, 0 to 28710
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   user_id  28711 non-null  int64
 1   book_id  28711 non-null  int64
dtypes: int64(2)
memory usage: 448.7 KB


**Question 1 :** 

How many books do not have an original title [books.csv] ? N

**Answer :** 39

In [11]:
books['book_id'].nunique() - books['original_title'].count()

39

In [12]:
# 2nd way 
books['original_title'].isna().sum()

39

**Removing null values from 'original_title'**

In [13]:
books_ = books.copy()
books_.isnull().sum()

book_id                       0
goodreads_book_id             0
best_book_id                  0
work_id                       0
books_count                   0
isbn                          4
isbn13                        0
authors                       0
original_publication_year     4
original_title               39
title                         0
language_code                80
average_rating                0
ratings_count                 0
work_ratings_count            0
work_text_reviews_count       0
ratings_1                     0
ratings_2                     0
ratings_3                     0
ratings_4                     0
ratings_5                     0
image_url                     0
small_image_url               0
NonEnglish                    0
dtype: int64

In [14]:
books_.dropna(subset=['original_title'],inplace=True)

In [15]:
books_.isnull().sum()

book_id                       0
goodreads_book_id             0
best_book_id                  0
work_id                       0
books_count                   0
isbn                          4
isbn13                        0
authors                       0
original_publication_year     2
original_title                0
title                         0
language_code                77
average_rating                0
ratings_count                 0
work_ratings_count            0
work_text_reviews_count       0
ratings_1                     0
ratings_2                     0
ratings_3                     0
ratings_4                     0
ratings_5                     0
image_url                     0
small_image_url               0
NonEnglish                    0
dtype: int64

**Question 2 :** 

How many unique books are present in the dataset ? Evaluate based on the 'book_id' [books.csv]  N

**Answer :** 556

In [16]:
books['book_id'].nunique()

556

**Question 3 :** 

How many unique users are present in the dataset [ratings.csv] ? N

**Answer :** 41797

In [17]:
ratings['user_id'].nunique()

41797

**Question 4 :**


Which book (title) has the maximum number of ratings based on ‘work_ratings_count’  [books.csv] ? S

**Answer :** The Lion, the Witch and the Wardrobe


In [18]:
books.sort_values(by='work_ratings_count', ascending=False).iloc[0,:]['original_title']

'The Lion, the Witch and the Wardrobe'

**Question 5 :** 

Which tag_id  is the most frequently used ie. mapped with the highest number of books [book_tags.csv]  ? (In case of more than one tag, mention the tag id with the least numerical value) N

**Answer :** 30574

In [19]:
book_tags.sort_values(by='count', ascending=False).iloc[0,:]['tag_id']

30574

In [20]:
# another way
book_tags.sort_values(by='count', ascending=False)

goodreads_book_id  tag_id   count
52100           11387515   30574  248197
7900               18131   30574  228455
700                 2998   30574  222156
32100             370493   30574  181787
9200               24213   30574  129449
...                  ...     ...     ...
49497            7278837    1174       1
44985            1693617   17170       1
44984            1693617   29816       1
50555            9366879   22689       1
45073            1728744   17085       1

[55800 rows x 3 columns]

**Question 6 :**

Which book (title) has the most number of counts of tags given by the user [book_tags.csv,books.csv]  ? S

**Answer :** Wonder

In [21]:
common_id = np.intersect1d(book_tags.columns ,books.columns)
common_id

array(['goodreads_book_id'], dtype=object)

In [22]:
book_tags.sort_values(by='count' ,ascending=False).iloc[0, :][common_id]

goodreads_book_id    11387515
Name: 52100, dtype: int64

In [23]:
books[books['goodreads_book_id'] == 11387515]['original_title']

18    Wonder
Name: original_title, dtype: object

**Question 7 :**

Which book (goodreads_book_id) is marked as to-read by most users [books.csv,toread.csv] ? N

**Answer :** Anastasia Krupnik

In [24]:
common_id = np.intersect1d(books.columns, toread.columns)
common_id

array(['book_id'], dtype=object)

In [25]:
book_id_val = toread.sort_values(by='book_id', ascending=False).iloc[0, :]['book_id']

book_id_val

9974

In [26]:
books[books.book_id==book_id_val]['original_title']

555    Anastasia Krupnik
Name: original_title, dtype: object

**Question 8 :**

Which is the least used tag, i.e. mapped with the lowest number of books [book_tags.csv]   ? (In case of more than one tag, mention the tag id with the least numerical value)  N

**Answer :** 5366

In [27]:
book_tags.sort_values(by='count').iloc[0,:]['tag_id']

5366

In [28]:
book_tags[book_tags['tag_id']==5366]['count']

46299    1
Name: count, dtype: int64

**Question 9 :**

Which book (title) has the minimum ‘average_rating’  [books.csv] ? S

**Answer :** The Chocolate War

In [29]:
books.sort_values(by='average_rating')['original_title']


250                                    The Chocolate War
399                                     The Whipping Boy
419                           Crispin: The Cross of Lead
392                                             Wildwood
148    Charlie and the Great Glass Elevator: The Furt...
                             ...                        
553                                                  NaN
336                                         The Gruffalo
501                            A Charlie Brown Christmas
206                                                  NaN
319       Disney's the Lion King (Disney Classic Series)
Name: original_title, Length: 556, dtype: object

**Question 10 :**

Which book (goodreads_book_id) has the least number of count of tags given by the user  [book_tags.csv,books.csv] ? N

**Answer :** Winnie the Pooh and Tigger Too

In [30]:
grd_id = book_tags.sort_values(by='count')['goodreads_book_id'].iloc[0]
grd_id

2617684

In [31]:
books[books.goodreads_book_id==grd_id]['original_title']

407    Winnie the Pooh and Tigger Too
Name: original_title, dtype: object

**Question 11 :**

How many tags are there in the dataset [book_tags.csv]  ? N

**Answer :** 3387

In [32]:
book_tags.tag_id.nunique()

3387

**Question 12 :**

What is the average rating of all the books in the dataset based on ‘average_rating’  [books.csv]  ? N

**Answer :** 4.092356115107914

In [33]:
books['average_rating'].mean()

4.092356115107914

**Question 13 :**

Find the number of books published in the year ‘2000’ based on the ‘original_publication_year’ [books.csv] ? N

**Answer :** 14 

In [34]:
len(books[books['original_publication_year'] == 2000])

14

**Question 14 :**

Predict sentiment using Textblob. How many positive titles (title) are there [books.csv] ? (cut-off >0) N (3 marks)

**Answer :** 48

In [35]:
from sklearn.metrics import classification_report
from textblob import TextBlob

In [39]:
books.isnull().sum()

book_id                       0
goodreads_book_id             0
best_book_id                  0
work_id                       0
books_count                   0
isbn                          4
isbn13                        0
authors                       0
original_publication_year     4
original_title               39
title                         0
language_code                80
average_rating                0
ratings_count                 0
work_ratings_count            0
work_text_reviews_count       0
ratings_1                     0
ratings_2                     0
ratings_3                     0
ratings_4                     0
ratings_5                     0
image_url                     0
small_image_url               0
NonEnglish                    0
dtype: int64

In [40]:
books_clear = books.copy()
books_clear.dropna(inplace=True)

In [41]:
books_clear['TextBlob_Subjectivity'] = books_clear['original_title'].apply(lambda x : TextBlob(x).sentiment.subjectivity)

books_clear['TextBlob_Polarity'] = books_clear['original_title'].apply(lambda x : TextBlob(x).sentiment.polarity)

In [42]:
books_clear['TextBlob_Analysis'] = books_clear['TextBlob_Polarity'].apply(lambda x : 'Positive' if x>0 else 'negative')

In [43]:
books_clear['TextBlob_Analysis'].value_counts()

negative    388
Positive     48
Name: TextBlob_Analysis, dtype: int64

**Question 15 :**

Plot a bar chart in Flourish with top 20 unique tags  in descending order of ‘user records’ (the number of users tagged the given tag_id with the goodreads_book_id) [book_tags.csv] and share the published link.
(2 marks)

**Answer :**

**Question 16 :**

Bucket the average_rating of books into 6 buckets [0,1,2,3,4,5] with 0.5 decimal rounding (eg: average_rating 3.5 to 4.4 will fall in bucket 4). 

Plot bar graph in Flourish to show total number of books in each rating bucket. [books.csv] and share the published link.
(2 marks)

**Answer :**